File: Complex/complex.cpp
#include <iostream>
#include <cmath>

#include "Complex.h"

using namespace std;

Complex::Complex(double real, double imaginary) {
    this->real = real;
    this->imaginary = imaginary;
}

Complex::~Complex() {}

double Complex::getReal() const {
    return real;
}

double Complex::getImaginary() const {
    return imaginary;
}

bool Complex::operator==(const Complex& other) const {
    return real == other.real && imaginary == other.imaginary;
}

bool Complex::operator!=(const Complex& other) const {
    return *this != other;
}

void Complex::operator+=(const Complex& other) {
    real += other.real;
    imaginary += other.real;
}

void Complex::operator-=(const Complex& other) {
    real -= other.real;
    imaginary -= other.real;
}

Complex Complex::operator+(const Complex& other) const {
    Complex result = *this;

    result.real += other.real;
    result.imaginary += other.imaginary;
    
    return result;
}


Complex Complex::operator-(const Complex& other) const {
    Complex result = *this;

    result.real -= other.real;
    result.imaginary -= other.imaginary;

    return result;
}

Complex Complex::operator*(const Complex& other) const {
    Complex result = *this;

    result.real *= other.real;
    result.imaginary *= other.imaginary;

    return result;
    }

std::ostream& operator<<(std::ostream& out, const Complex& complex) {
    if ( complex.getImaginary() < 0 ) {
        return out << complex.getReal() << complex.getImaginary();
    }

    return out << complex.getReal() << '+' << complex.getImaginary();
}

File: Java/Complex/Complex.java
import java.util.Vector;

/**
 * Created by andrew on 18.11.16.
 */
public class Complex implements Cloneable{
    private double real;
    private double imaginary;

    public Complex(double real, double imaginary) {
        this.real = real;
        this.imaginary = imaginary;
    }

    public Complex() {
        this(0.0, 0.0);
    }

    public double getReal() {
        return real;
    }

    public double getImaginary() {
        return imaginary;
    }

    public void setReal(double real) {
        this.real = real;
    }

    public void setImaginary(double imaginary) {
        this.imaginary = imaginary;
    }

    public Complex sum(Complex other) {
        return new Complex(real + other.real, imaginary + other.imaginary);
    }

    public Complex diff(Complex other) {
        return new Complex(real - other.real, imaginary - other.imaginary);
    }

    public Complex mult(Complex other) {
        return new Complex(real * other.real - imaginary * other.imaginary, real * other.imaginary + imaginary * other.real);
    }

    public boolean equals(Complex other) {
        return real == other.real && imaginary == other.imaginary;
    }

    public Complex clone() throws CloneNotSupportedException {
        return (Complex) super.clone();
    }

    public String toString() {
        if ( imaginary < 0 ) {
            return real + "" + imaginary + "i";
        }
        return real + " + " + imaginary + "i";
    }
}
